File: Workload/app/types/itemDefinition.ts
/**
 * TypeScript mirrors of FabInspector.Web/Workload/Contracts/*.cs.
 * Used by the React editors when reading/writing item definitions through
 * either the Fabric host (workloadClient.itemCrud) or the .NET backend.
 */

export interface ItemDefinitionPart {
    path: string;
    /** Always "InlineBase64" today. */
    payloadType: string;
    /** Base64-encoded UTF-8 JSON. */
    payload: string;
}

export interface ItemDefinition {
    parts: ItemDefinitionPart[];
}

export interface ItemDefinitionEnvelope {
    definition: ItemDefinition;
}

export interface RulesCatalogPointer {
    displayName: string;
    workspaceId: string;
    itemId: string;
    disabled: boolean;
}

export interface RulesCatalogPayload {
    /** Optional schema version per the toolkit backward-compat guidance. */
    version?: string;
    name: string;
    ruleSets: RulesCatalogPointer[];
}

/** Encode an arbitrary JSON-stringifiable payload as a single definition part. */
export function encodePart(path: string, payload: string): ItemDefinitionPart {
    return {
        path,
        payloadType: "InlineBase64",
        payload: btoa(unescape(encodeURIComponent(payload)))
    };
}

/** Decode a definition part back to its UTF-8 string contents. */
export function decodePart(part: ItemDefinitionPart): string {
    return decodeURIComponent(escape(atob(part.payload)));
}

export function findPart(envelope: ItemDefinitionEnvelope | null | undefined, path: string): ItemDefinitionPart | undefined {
    return envelope?.definition?.parts?.find(p => p.path.toLowerCase() === path.toLowerCase());
}

File: FabInspector.Web/Workload/Contracts/ItemDefinition.cs
using System.Text.Json.Serialization;

namespace FabInspector.Web.Workload.Contracts;

/// <summary>
/// Wire-format for a Fabric item definition as supplied by the Extensibility
/// Toolkit lifecycle endpoints. Mirrors the shape documented at
/// https://learn.microsoft.com/fabric/extensibility-toolkit/how-to-enable-remote-item-lifecycle
/// </summary>
public sealed class ItemDefinitionEnvelope
{
    [JsonPropertyName("displayName")]
    public string? DisplayName { get; set; }

    [JsonPropertyName("description")]
    public string? Description { get; set; }

    [JsonPropertyName("definition")]
    public ItemDefinition? Definition { get; set; }
}

public sealed class ItemDefinition
{
    [JsonPropertyName("parts")]
    public List<ItemDefinitionPart> Parts { get; set; } = new();
}

public sealed class ItemDefinitionPart
{
    [JsonPropertyName("path")]
    public string Path { get; set; } = string.Empty;

    /// <summary>Base64-encoded payload bytes.</summary>
    [JsonPropertyName("payload")]
    public string Payload { get; set; } = string.Empty;

    [JsonPropertyName("payloadType")]
    public string PayloadType { get; set; } = "InlineBase64";
}

public sealed class DeleteItemRequest
{
    [JsonPropertyName("deleteType")]
    public string DeleteType { get; set; } = "Hard"; // Hard | Soft
}
